AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/ClientBehavior/AWSCognitoAuthPlugin+UserBehavior.swift

```diff
@@ -70,7 +70,45 @@ public extension AWSCognitoAuthPlugin {
         fatalError("Not implemented")
     }
 
+    @available(*, deprecated, message: "Use async version of `getCurrentUser` instead")
     func getCurrentUser() -> AuthUser? {
-        fatalError("Not implemented")
+        var currentUser: AuthUser? = nil
+
+        // Note: blocking is not recommended
+        let group = DispatchGroup()
+
+        group.enter()
+        getCurrentUser { result in
+            currentUser = try? result.get()
+            group.leave()
+        }
+
+        group.wait()
+
+        return currentUser
     }
+
+    func getCurrentUser() async -> AuthUser? {
+        let authState = await authStateMachine.currentMachinState
+        if case .configured(let authenticationState, _) = authState,
+           case .signedIn(_, let signInData) = authenticationState {
+            let authUser = AWSCognitoAuthUser(username: signInData.userName, userId: signInData.userId)
+            return authUser
+        } else {
+            return nil
+        }
+    }
+
+    func getCurrentUser(closure: @escaping (Result<AuthUser?, Error>) -> Void) {
+        authStateMachine.getCurrentState { authState in
+            if case .configured(let authenticationState, _) = authState,
+               case .signedIn(_, let signInData) = authenticationState {
+                let authUser = AWSCognitoAuthUser(username: signInData.userName, userId: signInData.userId)
+                closure(.success(authUser))
+            } else {
+                closure(.success(nil))
+            }
+        }
```

Comment:
I discussed this with Chris. When a function uses `async` it operates in the cooperative thread pool which has behavior which may not behave if it is mixed with code which uses Dispatch on the caller side. By offering this version with the closure it will not use the cooperative thread pool and instead just use the operation queue which Auth is using. It is best to keep these code paths separate.